server/src/core/api-key/api-key.resolver.ts

```diff
@@ -0,0 +1,97 @@
+import { Args, Mutation, Query, Resolver } from '@nestjs/graphql';
+import { NotFoundException, UseGuards } from '@nestjs/common';
+
+import { accessibleBy } from '@casl/prisma';
+import { v4 } from 'uuid';
+
+import { AbilityGuard } from 'src/guards/ability.guard';
+import { AuthWorkspace } from 'src/decorators/auth-workspace.decorator';
+import { Workspace } from 'src/core/@generated/workspace/workspace.model';
+import { CreateOneApiKeyArgs } from 'src/core/@generated/api-key/create-one-api-key.args';
+import { ApiKey } from 'src/core/@generated/api-key/api-key.model';
+import { FindManyApiKeyArgs } from 'src/core/@generated/api-key/find-many-api-key.args';
+import { DeleteOneApiKeyArgs } from 'src/core/@generated/api-key/delete-one-api-key.args';
+import { CheckAbilities } from 'src/decorators/check-abilities.decorator';
+import {
+  CreateApiKeyAbilityHandler,
+  UpdateApiKeyAbilityHandler,
+  ReadApiKeyAbilityHandler,
+} from 'src/ability/handlers/api-key.ability-handler';
+import { JwtAuthGuard } from 'src/guards/jwt.auth.guard';
+import { UserAbility } from 'src/decorators/user-ability.decorator';
+import { AppAbility } from 'src/ability/ability.factory';
+import { TokenService } from 'src/core/auth/services/token.service';
+import { AuthToken } from 'src/core/auth/dto/token.entity';
+
+import { ApiKeyService } from './api-key.service';
+
+@UseGuards(JwtAuthGuard)
+@Resolver(() => ApiKey)
+export class ApiKeyResolver {
+  constructor(
+    private readonly apiKeyService: ApiKeyService,
+    private readonly tokenService: TokenService,
+  ) {}
+
+  @Mutation(() => AuthToken)
+  @UseGuards(AbilityGuard)
+  @CheckAbilities(CreateApiKeyAbilityHandler)
+  async createOneApiKey(
+    @Args() args: CreateOneApiKeyArgs,
+    @AuthWorkspace() { id: workspaceId }: Workspace,
+  ): Promise<AuthToken> {
+    const apiKeyId = v4();
+    const customApiKey = await this.tokenService.generateApiKeyToken(
```

Comment:
@martmull I think it's better to first create the entity in the database and then create the token, this way we can avoid generation of token that are not linked to the database if it fails.

Reply:
good point, will do